scripts/sbatch.py

```diff
@@ -0,0 +1,476 @@
+    # see above link for calculating formula
+    if (cpus_per_task is not SlurmHeader.cpus_per_task.default):
+        total_cpus = ntasks * cpus_per_task
+    else:
+        total_cpus = ntasks / ntasks_per_core
+    if shared == False:
+        nodes = ceil(total_cpus/cpu_per_node)
+        if ntasks_per_node is not SlurmHeader.ntasks_per_node.default:
+            nodes = nodes * (ntasks/ntasks_per_node)
+        total_cpus = nodes * cpu_per_node
+        total_mem  = nodes * mem_per_node
+    else:
+        total_mem = total_cpus * mem_per_cpu
+    #TODO:GPU may need to be included
+    requested_hours = (total_seconds_requested * (total_cpus * cpu_billing_weights)) + (
+        total_mem * mem_billing_weights
+    ) 
+    return requested_hours
+
+
+def get_available_chours(json_file, account, username):
+    """
+       Read the core hours json file that's populated from the dashboard.db and
+       return the available core hours for a specified user in specified account
+    """
+    with open(json_file, "r") as f:
+        json_array = json.load(f)
+    for d in json_array:
+        if d.get(account) is not None:
+            for user_dict in d[account]:
+                if user_dict.get(username) is not None:
+                    return (
+                        user_dict[username]["allocation"] - user_dict[username]["used"]
+                    )
+            #no data for a user is found while account exist
+            sys.exit(
+                "No core hours usage info for {} {} from json file {}".format(
+                    account, username, json_file))
+    #all project names looped raise error
+    raise IndexError("No account info for {} from json file {}".format(account, json_file))
+
+
+def process_slurm_header(sl_file):
+    """Read a slurm file and returns the wanted header values as a dictionary"""
+    with open(sl_file) as f:
+        lines = f.readlines()
+    #example line: SBATCH --header=value
+    header_dict = {h.value: h.default for h in SlurmHeader}
+    for line in lines:
+        if line.upper().startswith("#SBATCH"):
+            for header in SlurmHeader:
+                if header.type is not bool and "--{}=".format(header.id) in line:
+                    value = line.strip().split("--{}=".format(header.id))[-1]
+                    #special case for WCT(time) parameter
+                    if header is SlurmHeader.time:
+                        #a day is specified, i.e. "%d-%H:%M:%S"
+                        wct_list = list( map(int, re.findall(r"[-+]?\d*\.\d+|\d+", value)))
+                        #filter the value to contain 4 fields, day, h, m, s
+                        while len(wct_list) < 4: 
+                            wct_list.insert(0,0)
+                        value = datetime.timedelta(days=wct_list[0],hours=wct_list[1],minutes=wct_list[2],seconds=wct_list[3]).total_seconds()
+                    header_dict[header.value] = header.type(value)
+                    break
```

Comment:
doesn't actually do anything.

Reply:
since the its looping through all potential header, it breaks when a match is found?

Reply:
Yep. That's right carry on.